units/commit_queue.go

```diff
@@ -0,0 +1,352 @@
+package units
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/model/commitqueue"
+	"github.com/evergreen-ci/evergreen/model/event"
+	"github.com/evergreen-ci/evergreen/model/patch"
+	"github.com/evergreen-ci/evergreen/model/user"
+	"github.com/evergreen-ci/evergreen/thirdparty"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/google/go-github/github"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/level"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	commitQueueJobName = "commit-queue"
+)
+
+func init() {
+	registry.AddJobType(commitQueueJobName, func() amboy.Job { return makeCommitQueueJob() })
+}
+
+type commitQueueJob struct {
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+	QueueID  string `bson:"queue_id" json:"queue_id" yaml:"queue_id"`
+}
+
+func makeCommitQueueJob() *commitQueueJob {
+	job := &commitQueueJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    commitQueueJobName,
+				Version: 0,
+			},
+		},
+	}
+	job.SetDependency(dependency.NewAlways())
+
+	return job
+}
+
+func NewCommitQueueJob(queueID string, id string) amboy.Job {
+	job := makeCommitQueueJob()
+	job.QueueID = queueID
+	job.SetID(fmt.Sprintf("%s:%s_%s", commitQueueJobName, queueID, id))
+
+	return job
+}
+
+func (j *commitQueueJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	projectRef, err := model.FindOneProjectRef(j.QueueID)
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "can't find project for queue id %s", j.QueueID))
+		return
+	}
+
+	if !projectRef.CommitQEnabled {
+		return
+	}
+
+	cq, err := commitqueue.FindOneId(j.QueueID)
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "can't find commit queue for id %s", j.QueueID))
+	}
+	nextItem := cq.Next()
+	nextItemInt, err := strconv.Atoi(nextItem)
+	if err != nil {
+		j.AddError(errors.Wrapf(err, "can't parse next item \"%s\" as int", nextItem))
+		return
+	}
+
+	alreadyProcessed, err := versionExists(nextItemInt)
```

Comment:
So when one job has already picked up the head of the queue but hasn't finished merging yet another job will stop here (the front of the queue is only popped when the merge is attempted).
I also considered adding a field to the commit queue to indicate the front of the queue is in progress, setting that here and clearing it when we pop the queue. This would have an advantage that were another job to start at the same time as this one and both try to create a version for this item at around the same time we could have two versions for the same item and try merging it twice. I decided not to do this since:
- it's cleaner this way
- the jobs are started in regular intervals

Reply:
Does this approach handle races when github's API is slow? Or more generally, what if a commit queue job for a specific project overruns the one before it?

Reply:
No.
I could put this check right in front of the attempt to create a version (line 128), which could help for GitHub being slow, but the only way I could think of to eliminate the possibility of a race is what I mentioned to mark the queue as busy.

Reply:
I think what you have right now is sufficient. I'd like to avoid having a distributed lock here